runtime/freelist.c

```diff
@@ -534,60 +348,1395 @@ header_t *caml_fl_merge_block (value bp)
+      flp [flp_size++] = ff_last;
+    }
+  }else{
+    value prev;
+
+    prev = Ff_head;
+    cur = Next_small (prev);
+    while (cur != Val_NULL && Bp_val (cur) < Bp_val (bp)){
+      CAMLassert (Bp_val (prev) < Bp_val (bp) || prev == Ff_head);
+      /* XXX TODO: extend flp on the fly */
+      prev = cur;
+      cur = Next_small (prev);
+    }
+    CAMLassert (Bp_val (prev) < Bp_val (bp) || prev == Ff_head);
+    CAMLassert (Bp_val (cur) > Bp_val (bp) || cur == Val_NULL);
+    Next_small (Field (bp, 1)) = cur;
+    Next_small (prev) = bp;
+    /* When inserting blocks between [caml_fl_merge] and [caml_gc_sweep_hp],
+       we must advance [caml_fl_merge] to the new block, so that [caml_fl_merge]
+       is always the last free-list block before [caml_gc_sweep_hp]. */
+    if (prev == caml_fl_merge && (char *) bp < caml_gc_sweep_hp){
+      caml_fl_merge = Field (bp, 1);
+    }
+    ff_truncate_flp (bp);
+  }
+}
+
+static void ff_make_free_blocks
+  (value *p, mlsize_t size, int do_merge, int color)
+{
+  mlsize_t sz;
+
+  while (size > 0){
+    if (size > Whsize_wosize (Max_wosize)){
+      sz = Whsize_wosize (Max_wosize);
+    }else{
+      sz = size;
+    }
+    *(header_t *)p = Make_header (Wosize_whsize (sz), 0, color);
+    if (do_merge) ff_merge_block (Val_hp (p), NULL);
+    size -= sz;
+    p += sz;
+  }
+}
+
+/********************* best-fit allocation policy *********************/
+
+/* quick-fit + FIFO-ordered best fit (Wilson's nomenclature)
+   We use Standish's data structure (a tree of doubly-linked lists)
+   with a splay tree (Sleator & Tarjan).
+*/
+
+/* [BF_NUM_SMALL] must be at least 4 for this code to work
+   and at least 5 for good performance on typical OCaml programs.
+   For portability reasons, BF_NUM_SMALL cannot be more than 32.
+*/
+#define BF_NUM_SMALL 16
+
+/* How many blocks to pre-allocate at once for a small free list when
+   it and all the larger ones are empty.
+   100 should be a good number to amortize the overhead. */
+#define BF_PREALLOC_NUM 100
+
+static struct {
+  value free;
+  value *merge;
+} bf_small_fl [BF_NUM_SMALL + 1];
+int bf_small_map = 0;
+
+/* Small free blocks have only one pointer to the next block.
+   Large free blocks have 5 fields:
+   tree fields:
+     - node flag
+     - left son
+     - right son
+   list fields:
+     - next
+     - prev
+*/
+typedef struct large_free_block {
+  int isnode;
+  struct large_free_block *left;
+  struct large_free_block *right;
+  struct large_free_block *prev;
+  struct large_free_block *next;
+} large_free_block;
+
+static inline mlsize_t bf_large_wosize (struct large_free_block *n) {
+  return Wosize_val((value)(n));
+}
+
+static struct large_free_block *bf_large_tree;
+static struct large_free_block *bf_large_least;
+/* [bf_large_least] is either NULL or a pointer to the smallest (leftmost)
+   block in the tree. In this latter case, the block must be alone in its
+   doubly-linked list (i.e. have [isnode] true and [prev] and [next] NULL)
```

Comment:
I think `prev` and `next` point back to the list node when the list is a singleton, rather than being `NULL`.